File: day_5/solution2.py
# read input
def read_input(file_path: str) -> tuple[list[tuple[int, int]], list[int]]:
    fresh_id_range = []
    available_id = []
    flag = False
    with open(file_path, "r") as file:
        for line in file:
            line = line.strip()
            if line == "":
                flag = True
                continue
            if flag:
                available_id.append(int(line))
            else:
                temp = line.split("-")
                fresh_id_range.append((int(temp[0]), int(temp[1])))
    return fresh_id_range, available_id


def find_fresh_ids(fresh_id_range: list[tuple[int, int]]) -> int:
    result = 0
    sorted_ranges = sorted(fresh_id_range, key=lambda x: x[0])
    new_ranges = [sorted_ranges[0]]
    for start, end in sorted_ranges[1:]:
        last_start, last_end = new_ranges[-1]
        if start <= last_end + 1:
            new_ranges[-1] = (last_start, max(last_end, end))
        else:
            new_ranges.append((start, end))
    for start, end in new_ranges:
        result += end - start + 1
    return result


path = "./input.txt"
# path = "./input_test.txt"

fresh_id_range, available_id = read_input(path)

result = find_fresh_ids(fresh_id_range)
print(f"Result: {result}")

File: day_5/solution_rust/src/solution_2.rs
use std::fs::File;
use std::io::{BufRead, BufReader};

type Pair = (i64, i64);
type PairList = Vec<Pair>;
type ValueList = Vec<i64>;
type IoResult<T> = Result<T, std::io::Error>;

fn read_input(file_path: &str) -> IoResult<(PairList, ValueList)> {
    let file = File::open(file_path)?;
    let reader = BufReader::new(file);
    let mut fresh_id_range: Vec<(i64, i64)> = Vec::new();
    let mut available_id: Vec<i64> = Vec::new();
    let mut flag = false;
    for line in reader.lines() {
        let temp = line?.trim().to_string();
        if temp.is_empty() {
            flag = true;
            continue;
        }
        if !flag {
            fresh_id_range.push((
                temp.split('-').next().unwrap().parse::<i64>().unwrap(),
                temp.split('-').nth(1).unwrap().parse::<i64>().unwrap(),
            ));
        } else {
            available_id.push(temp.parse::<i64>().unwrap());
        }
    }
    Ok((fresh_id_range, available_id))
}

fn find_fresh_ids(fresh_id_range: &mut [(i64, i64)]) -> i64 {
    let mut count = 0;
    fresh_id_range.sort_by_key(|range| range.0);
    let mut new_ranges: Vec<(i64, i64)> = Vec::new();
    new_ranges.push(fresh_id_range[0]);
    for &(start, end) in fresh_id_range.iter().skip(1) {
        let last_range = new_ranges.last_mut().unwrap();
        let (_, last_end) = *last_range;
        if start <= last_end + 1 {
            // Extend the last range if needed
            if end > last_end {
                last_range.1 = end;
            }
        } else {
            // Disjoint range: start a new one
            new_ranges.push((start, end));
        }
    }
    for (start, end) in new_ranges {
        count += end - start + 1;
    }
    count
}

pub fn run() {
    // let path = "../input_test.txt";
    let path = "../input.txt";
    let (mut fresh_id_range, _) = match read_input(path) {
        Ok((fresh_id_range, available_id)) => (fresh_id_range, available_id),
        Err(e) => {
            eprintln!("Error reading input: {}", e);
            return;
        }
    };
    let result = find_fresh_ids(&mut fresh_id_range);
    println!("Number of fresh IDs: {}", result);
}
